packages/nitro-protocol/contracts/examples/EmbeddedApplication.sol

```diff
@@ -0,0 +1,438 @@
+        );
+
+        // ensure A,B part of the outcome of X has been absorbed into the outcome of J
+        require(
+            Xallocation[0].amount == toAllocation[0].amount &&
+                Xallocation[1].amount == toAllocation[1].amount &&
+                Xallocation[0].destination == toAllocation[0].destination &&
+                Xallocation[1].destination == toAllocation[1].destination,
+            'X / J outcome mismatch'
+        );
+        return true;
+    }
+
+    function validateSupportProofForX(AppData memory fromAppData, AppData memory toAppData)
+        internal
+        pure
+        returns (Outcome.AllocationItem[] memory Xallocation)
+    {
+        // The following checks follow the protocol-level validTransition function
+        // They are the members of FixedPart that do not affect the channelId
+        require(
+            fromAppData.supportProofForX.fixedPart.appDefinition ==
+                toAppData.supportProofForX.fixedPart.appDefinition,
+            'X.appDefinition changed'
+        );
+        require(
+            fromAppData.supportProofForX.fixedPart.challengeDuration ==
+                toAppData.supportProofForX.fixedPart.challengeDuration,
+            'X.challengeDuration changed'
+        ); // this is actually never used so could be 0
+
+        // A support proof requires 2 signatures
+        // But it may have either 1 or two states.
+        // If both signatures on the same state, it is supported
+        // Else one signature per state and we must check for a valid transition
+        require(
+            toAppData.supportProofForX.variableParts.length == 1 ||
+                toAppData.supportProofForX.variableParts.length == 2,
+            '1 or 2 states required'
+        );
+
+        bytes32 appPartHash = keccak256(
+            abi.encode(
+                toAppData.supportProofForX.fixedPart.challengeDuration,
+                toAppData.supportProofForX.fixedPart.appDefinition,
+                toAppData.supportProofForX.variableParts[0].appData
+            )
+        );
+
+        // hash the greatest state first (either the later of a pair, or the only state provided)
+        bytes32 greaterStateHash = keccak256(
+            abi.encode(
+                IForceMove.State(
+                    toAppData.supportProofForX.turnNumTo,
+                    false, // Assume isFinal is false
+                    toAppData.channelIdForX,
+                    appPartHash,
+                    keccak256(
+                        toAppData.supportProofForX.variableParts[toAppData
+                            .supportProofForX
+                            .variableParts
+                            .length - 1]
+                            .outcome
+                    )
+                )
+            )
+        );
+
+        if (
+            (toAppData.supportProofForX.whoSignedWhat[0] == 0) &&
+            (toAppData.supportProofForX.whoSignedWhat[1] == 0)
```

Comment:
`whoSignedWhat` seems like a confusing name -- I've been confused by it before. 
```
        uint8[] memory whoSignedWhat // whoSignedWhat[i] is the index of the state in stateHashes that was signed by participants[i]
```

I think it's confusing because `whoSignedWhat` is actually a list of "_what_ is signed, indexed by _whom_".

Brainstorming: Is something more along the lines of `didSignState` more clear? `didSignState[2] == 3` means "participant 2 did sign state 3". (I don't like `didSignState` either, just throwing an idea out there.)

Reply:
maybe `stateIdxSignedBy`? `stateIdxSignedBy[2] == 1` => "participant 2 signed the state with index 1"

Reply:
or `stateIndexSignedByParticipant`?

Reply:
We should make a new issue for this and tackle it separately to this PR.

Reply:
#3605 3605